scripts/utilities/datarun.py

```diff
@@ -0,0 +1,360 @@
+                
+        # create a run with the given parameters
+        design = rescal_utilities.Design_a_run()
+        design.set_parameters(self.parameters)
+
+        # find file paths that are in the real data folder and fix them
+        # using self.real_data_directory 
+        design.parameters.set_real_data_path(self.real_data_directory)
+            
+        # write the par file
+        design.parameters.write(self.par)
+
+        # run genesis if neccesary
+        if 'premade_csp' not in self.parameters.keys():
+            # run genesis and put the output where specified by 'Csp_file' parameter
+            p = subprocess.Popen([self.genesis_executable, '-f',  self.par, '-s', str(2000)],
+                                 stdout=subprocess.DEVNULL)
+            p.wait()
+
+        
+        # now get the cmd line args for rescal
+        cmd_args = design.run_script.rescal_call_args()
+
+        # put rescal and the par file in, after nice if nice exists
+        if cmd_args[0].strip() == 'nice':
+            cmd_args = cmd_args[0] + [self.rescal_executable, self.par] + cmd_args[1:]
+        else:
+            cmd_args = [self.rescal_executable, self.par] + cmd_args
+        self.rescal_args = cmd_args
+        
+            
+    def receive_process_data(self):
+        '''Get the data size, which will be in a 4 byte integer.
+        Get the data itself. Create a cellspace.CellSpace to
+        process the data and keep the height_maps and ffts 
+        based on flags.'''
+        
+        data_size_bytes = os.read(self.r, 4)
+        if not data_size_bytes:
+            return
+        data_size = struct.unpack('i', data_size_bytes)[0]
+
+        self.data = os.read(self.r, data_size)
+        
+        c = cellspace.CellSpace(self.data)
+        # get the actual heightmap and ffts arrays out
+        if self.keep_height_maps:
+            self.height_maps.append(c.height_map.height_map)
+        if self.keep_ffts:
+            self.ffts.append(c.height_map.fft_blur)
+
+        
+    def write_meta_data(self):
+        '''Write the rescal parameters and execuation call to a file
+        in the experiment directory.'''
+        with open(self.meta_data, 'w+') as f:
+            f.write('parameters = ' + str(self.parameters) + '\n')
+            f.write('exec = ' + str(self.rescal_args) + '\n')
+
+
+    def run_simulation(self):
+        '''Run ReSCAL and get the cellspace data
+        that would otherwise be written to a file. Process and save the data
+        to files and write out the meta-data.'''
+
+        # setup pipe to rescal
+        r,w = os.pipe()
+
+        self.r = r
+        
+        # allow rescal to inherit the write side of the pipe
+        os.set_inheritable(w, True)
+
+        # add the pipe to self.rescal args
+        self.rescal_args = self.rescal_args + ['-data_pipe',  str(w)]
+
+        # start a rescal process that can send data back
+        rescal = subprocess.Popen(self.rescal_args,
```

Comment:
This section seems convoluted. Can you add comments explaining why it is done this way or switch ti a simpler formulation (`check_output`)?

Reply:
I'm using a pipe which gives me a new file descriptor because I don't want to use the stdout of ReSCAL because it's full of text I don't want to deal with. I'm telling ReSCAL which file descriptor to use with a command line argument. I think all the convenient methods for this use stdin and stdout.

Reply:
I'll add comments more comments on this in the code.

Reply:
If the file descriptor has a number, it looks like you can pass that to `check_output`:
> stdin, stdout and stderr specify the executed program’s standard input, standard output and standard error file handles, respectively. Valid values are PIPE, an existing file descriptor (a positive integer), an existing file object, and None

From [here](https://docs.python.org/2.7/library/subprocess.html#frequently-used-arguments).

Reply:
I still don't think this is the behavior I want. The python3 documentation says that check_ouput is equivalent to 
run(..., check=True, stdout=PIPE).stdout
Which means that I'll get everything that's sent to stdout.
Also, I don't want all the data at once in one big chunk at the end of the run, I want it one .csp file at a time which is processed while ReSCAL is generating the next .csp file.